ENH: Add Lomb-Scargle Periodogram

jakevdp opened the pull request:
The PR contains six complementary implementations of the Lomb-Scargle periodogram, along with full unit tests and documentation. It is adapted from http://github.com/jakevdp/lombscargle/ and from http://astroml.org. 

For benchmarks of the approaches, see [this notebook](https://gist.github.com/jakevdp/86c20012149e4fe2f7c656145c893df6).

A build of the docs from this PR can be found [here](http://jakevdp.github.io/astropy/stats/lombscargle.html).

Also – I know this is a rather large contribution and adds to the maintenance challenge of the astropy project. If you think this would better serve the community as a standalone affiliated ``lombscargle`` package, I'd be happy to go that route as well!

jakevdp commented:
Also – I know this is a rather large contribution and adds to the maintenance challenge of the astropy project. If you think this would better serve the community as a standalone affiliated ``lombscargle`` package, I'd be happy to go that route as well!

jakevdp commented:
Test failures due to numpy 1.7 in some cases, and lack of scipy in some cases. Let me try to fix it.

jakevdp commented:
OK – I skip appropriate tests when numpy 1.8+ and scipy are not available; also still test the ``'auto'`` method to make sure it doesn't select an invalid approach in these cases. Hopefully this will pass!

jakevdp commented:
Remaining failures are on NumPy 1.7: it's a doctest for a utility that is never called unless NumPy is 1.8 or newer. Is there any directive to conditionally skip a doctest block based on a package version? I suppose I could just remove the whole block, but it's actually a quite useful example.

I'd be open to suggestions.

pllim commented:
I have to admit I don't know what "ENH" means...

@larrybradley , are you interested to have a look?

bsipocz commented:
@pllim - enhancement. Many packages have stricter (and actually quite useful) PR naming requirements, so you can see straight away from the name that what type of PR it is.

Thanks @jakevdp, this is a super great addition.

jakevdp commented:
Anyone know why Travis and coveralls failed to run on the last commit?

bsipocz commented:
@jakevdp - I never saw this happening before, as there cannot be any conflicts as this is right at the top of master. 

Maybe repushing the latest commit would start it up? (I think minor amend to the commit message and force push would do the trick).

pllim commented:
@bsipocz , we also have "Enhancement" tag. Should we remove that tag in favor of this "ENH" acronym?

pllim commented:
I think we should merge #4784 soon, which would make the Numpy 1.6 failure here irrelevant.

jakevdp commented:
I also have a Cython implementation that I might think about adding here.

bsipocz commented:
@pllim - Why not have both? Tags are useful to filter with one click and this is in the overview and e-mail notifications.

@jakevdp - stats maintainers should comment on this, but I for one, would definitely switch to use this astropy LS rather than python wrapped third party C code (e.g. vartools) if it has comparable speed. So +1 for the Cython.

jakevdp commented:
@bsipocz - I think the Python implementation here will already be comparable to the ``fasper`` code used in vartools: it is the same approach, and all significant computations are pushed into C via numpy.

bsipocz commented:
@jakevdp - Ohh, that's great to hear. Thanks.

jakevdp commented:
The cython code I have is equivalent to ``method='slow'`` here, but much faster... it's only really relevant for smaller problems when the approximations used by ``method='fast'`` are not warranted

(The default ``method='auto'`` does all this logic for you)

jakevdp commented:
Added a cython implementation – if you're curious I put together some benchmarks [here](https://gist.github.com/jakevdp/86c20012149e4fe2f7c656145c893df6)

jakevdp commented:
Failure in coveralls seems unrelated to this PR.

pllim commented:
I have a hunch that rebasing to latest master might solve the coverage failure.

jakevdp commented:
Rebased on master

jakevdp commented:
For convenience I put up a build of the docs from this PR; Lomb Scargle documentation is [here](http://jakevdp.github.io/astropy/stats/lombscargle.html).

crawfordsm commented:
thanks @jakevdp it is useful to see the docs

pllim commented:
The coverage is only complaining about WCS C-code now, which does not seem to be related to this PR as well.

pllim commented:
This looks very nice if we can fit into 1.2 :smile:

jakevdp commented:
The last few commits have been mostly minor documentation & cleanup stuff. I think the PR is ready for review, if anyone is up for it!

pllim commented:
Just heads up -- When milestone is decided, you will also need to add a change log. And ideally squash the commits before merge. Thanks!

mirca commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,256 @@
+        error or sequence of observational errors associated with times t
+    fit_bias : bool (optional, default=True)
+        if True, include a constant offet as part of the model at each
+        frequency. This can lead to more accurate results, especially in then
```
especially in the*

jakevdp replied:
fixed – thanks.

adrn commented:
FWIW: My opinion is that this belongs in an affiliated package. In fact, I think all of `astropy.stats` belongs in an affiliated package, and this could be one subpackage within that. Right now, `astropy.stats` is just a repository for dumping random statistical utilities and, I would argue, is still somewhat exploratory. That to me sounds like one of the visions for an affiliated package: a place to add / try out new features before merging in to astropy core as a more mature subpackage. It could end up going the other direction (start in astropy core -> affiliated package) if it is developed enough. Maybe that's ok? Anyways, just expressing my thoughts!

pllim commented:
@adrn , I think the discussion of moving the entire `stats` subpackage out of Astropy is a separate issue unrelated to this PR (PiA 2017 unconference?). Given that `stats` is here to stay (at least for now), do you agree that this belongs in `stats` or otherwise?

adrn commented:
@pllim I agree that evicting `astropy.stats` is a larger issue, but worth thinking about for this PR. I think that this would belong in whatever `astropy.stats` becomes, but do we really want to merge this in to core, have people start writing code that `from astropy.stats import lombscargle` and then have to deal with deprecationwarnings and etc. if we later decide to make it a separate package? I think this will be a very popular / heavily used addition! The answer may be "yes," but I just want us to think about it!

pllim commented:
@adrn , people already use other things from `stats` (sigma clipping, etc). So regardless whether this is merged or not, we have that "migration" problem anyway. Whether if I move 10 couches or 11 couches from my house to another house, it is still the same problem.

astrofrog commented:
I would argue that this should go directly into the core. If people have opinions about how the astropy.stats package should be organized, or if some of the names here are not optimal, we can just discuss it now. I don't think we've ever had issues with old astropy.stats functions that needed significant changes?

jakevdp commented:
Note the current CI failures are unrelated to this branch: a URL read error in Travis, and a coverage error in a module I've not touched.

pllim commented:
Yeah, I wouldn't worry about the failure. A rebase to master would probably fix the coverage.

crawfordsm commented:
@adrn My feeling is as the stats package is currently constructed (a place for statistical tools used by astronomers) this fits perfectly into that definition so it should go into the stats package.   I'd agree that it would be useful to have a larger scale discussion about it though but probably best outside this PR

crawfordsm commented:
@jakevdp I can possible provide a final review of this next week, but would have anyone suggestions for possible  scientific reviewers of its performance and description?

adrn commented:
Not debating the usefulness of this code and totally agree that it belongs in *some* stats package! Happy for this to be merged and to move discussion elsewhere...

pllim commented:
That discussion can go to #4846.

mhvk commented:
I can see discussions about whether or not many things in the `stats` package should be in core, but to me, this particular contribution seems very much at the level where we would like everything else to be!
I will try to comment in more detail on the implementation...

astrofrog commented:
@mhvk - do you still want to review on this?

We should make sure this gets merged before Friday! :)

mwcraig commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,256 @@
+    array([ 0.0204304 ,  0.01393845,  0.35552682,  0.01358029,  0.03083737])
+
+    If the inputs are astropy Quantities with units, the units will be
+    validated and the outputs will also be Quantities with appropriate units:
```
Is the statement about validation correct? Line 67 indicates it isn't...

jakevdp replied:
That comment in line 67 is out of date. I'll remove it. The current code handles units correctly, and it's all tested in ``astropy/stats/lombscargle/tests/test_lombscargle.py``

mwcraig commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,256 @@
+
+        method_kwds : dict (optional)
+            additional keywords to pass to the lomb-scargle method
+        normalization : string (optional, default='normalized')
```
Unless the intent is to offer other normalization options in the future, could this simply be a boolean?

jakevdp replied:
There are a *lot* of different normalizations of the periodogram used in the literature. I was going to dig through and implement others, but I didn't have the time. I thought I'd leave it as a string in case people want to add more in the future.

jakevdp replied:
[Baluev 2007](http://arxiv.org/abs/0711.0330) lists a few of them. Do you think it would be worth implementing those?

mwcraig replied:
I don't think there is any need to implement them now, and since there are more coming in the future the current approach makes sense. Might be useful to open an issue about adding other normalizations as a reminder once this is merged (having an open issue for it also makes it easier for potential new contributors to come across it).

mhvk replied:
I'd have a (very slight) preference for no normalisation being `None`.  But very slight, so feel free to ignore.

mwcraig commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,256 @@
+                            assume_regular_frequency=True)
+        return frequency, power
+
+    def power(self, frequency, normalization='normalized', method='auto',
```
Suggestion: merge `power` and `autopower` into a single function, `power`, with a boolean argument `autofrequency`, returning just the power if `autofrequency` is false, and both frequency and power otherwise.

Requires a little more logic in the method, but cuts several lines of otherwise identical docstring and provides an (overall) simpler interface, I think.

jakevdp replied:
Initially I had it that way. The problem is that ``autopower()`` must return a frequency, while for ``power()`` it is redundant to return the frequency, because the user must provide it.

I don't like interfaces where the number of return values depends on how you call the function, so I chose to split this into two methods.

Would be open to hearing other opinions on this.

MSeifert04 replied:
Just thinking aloud: You could always return frequency and power. In case of `power()` it would mean just returning the input but that's better then input-based-output-length. Or does that complicate things somewhere else? I must admit I haven't checked the internals.

If it's just docstrings, I've added a [`format_doc`](http://astropy.readthedocs.io/en/latest/api/astropy.utils.decorators.format_doc.html#astropy.utils.decorators.format_doc) decorator that can be used if similiar functions have only slightly different docstrings. (I realize the documentation is not very well formulated ...) and there are some similar (but different) decorators in astropy-helpers somewhere...

mwcraig replied:
Yeah, I'm not real fond of a variable number of return values either...was looking for a way to reduce duplication.

mhvk replied:
I think I like @MSeifert04's suggestion best, but with the variation that for `power` you have the frequencies calculated if they are not passed in (i.e., default is `None`), and that, therefore, `autopower` can be removed altogether.  Then, if you do want to pass in frequencies, you might do either `_, power = ls.power(freq)` or `freq, power = ls.power(freq)`.

jakevdp replied:
I strongly disagree. Adding a needlessly awkward API just so we can avoid a few lines of duplicated code seems like putting the cart before the horse.

mhvk replied:
It depends a bit on whether the "standard" usage would be `autopower`; if both are used about as often, then fine to just keep as is.

mwcraig commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,255 @@
+        y : np.ndarray, length n_samples
+            The model fit corresponding to the input times
+        """
+        # TODO: handle units correctly
```
Another stray `TODO`?

jakevdp replied:
No, this function actually does not handle units yet. It's not really an essential feature though...

mhvk replied:
We should support quantities, at least eventually! It turns out to be very helpful to not have to be sure that, e.g., the times and frequency one specifies have to be in complementary units.  Some comments in `periodic_fit`.

mwcraig commented on astropy/stats/lombscargle/implementations/fastchi2_impl.py:
```diff
@@ -0,0 +1,123 @@
+        parameters describing the frequency grid, f = f0 + df * arange(Nf).
+    normalization : string (optional, default='normalized')
+        Normalization to use for the periodogram
+        TODO: figure out what options to use
```
Another TODO...

jakevdp replied:
Thought I'd removed that one

mwcraig commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+The ``lombscargle`` function here is essentially a sophisticated switch
+statement for the various implementations available in this submodule
+"""
+__all__ = ['lombscargle', 'available_methods']
```
Please add `from __future__ import print_function, division`

mwcraig commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+"""Tools for maximum likelihood estimation associated with Lomb-Scargle"""
```
Please add `from __future__ import print_function, division`

mwcraig commented on astropy/stats/lombscargle/implementations/tests/test_lombscargle.py:
```diff
@@ -0,0 +1,139 @@
+import pytest
```
I think the preference is to use the pytest bundled with astropy for the testing, so this would be:

`from ......tests.helper import pytest`

It is highly likely I've miscounted the number of `.`s needed...

mwcraig commented on astropy/stats/lombscargle/implementations/tests/test_mle.py:
```diff
@@ -0,0 +1,54 @@
+import pytest
```
See comment at the top of `test_lombscargle.py`

mwcraig commented on astropy/stats/lombscargle/implementations/tests/test_utils.py:
```diff
@@ -0,0 +1,83 @@
+
+import numpy as np
+from numpy.testing import assert_allclose, assert_equal
+import pytest
```
Please see `test_lombscargle.py`

mwcraig commented on astropy/stats/lombscargle/tests/test_lombscargle.py:
```diff
@@ -0,0 +1,161 @@
+import pytest
```
pytest import again...

mwcraig commented:
Overall, this looks great! Wasn't sure how much of the `gatspy` content was going to end up in here, but this seems to capture the core (pun intended) of the generally-useful algorithms and provides a convenient, uniform interface.

Unfortunately, I don't have the background to review the details of each of the implementations in the pull request, but after reading through the code and docs have a few questions/suggestions in addition to mostly minor inline comments:

+ The `assume_regular_frequency` option shows up in a few places. I'm having trouble imagining cases where setting it to `True` would be the right thing to do. Irregular frequencies should be allowed, of course, for the methods that don't require regular frequencies but it seems like the ones that do should always check that the requirement is satisfied.

+ Although it is made clearly in several places in docstrings, I don't think the narrative documents explicitly say that the frequencies should not be angular frequencies. The narrative docs do have examples that refer to frequencies in cycles per sec, so maybe it isn't absolutely necessary, but it would be helpful to be explicit about that.

+ Some advice as to when (if ever) to use `center_data` and when to use `fit_bias` would be helpful. IIRC, in either your multi-band periodogram paper or the gatspy docs you make the point that subtracting the mean (i.e. `center_data`) can give you a biased result if the sampling is biased towards one part of the signal (e.g. you only observe a variable near its maximum, or you detect an object only during the brighter part of its variation) unless you also fit for the bias. If `fit_bias` is `False`, then I think you must center the data (I know the scipy implementation requires it, at least), and if `fit_bias` is `True` then you may as well center the data (as you currently do), so perhaps `center_data` could be removed altogether?

+ I think typically the preference when checking validity of parameters to functions/methods the astropy practice has been to print an explicit error message; in several places you have `assert` statements instead. Not sure that is urgent to fix (and it may be I'm wrong about the practice).

mhvk commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,255 @@
+"""Main Lomb-Scargle Implementation"""
```
Usually in astropy, we do not have hidden files, though we do use the underscore for indicating variables that are not public. By analogy of `time` and `units`, perhaps call this file `core.py`?

larrybradley replied:
+1 for `core.py`

jakevdp replied:
Will do.

mhvk commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+    if dy is None:
+        dy = 1
+
+    t, y, dy = np.broadcast_arrays(t, y, dy)
```
To preserve quantities, you should add `subok=True`. But this works only for newer versions of numpy, so at the top import the backported version and use that:
```
from ....astropy.utils.compat.numpy import broadcast_arrays
```
(please check number of dots...)

mhvk replied:
Actually, why is the broadcast necessary?

jakevdp replied:
``dy`` can be a scalar, and this is the easiest way to make sure all the inputs are the same shape and raise an appropriate error if they are not.

mhvk commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+        dy = 1
+
+    t, y, dy = np.broadcast_arrays(t, y, dy)
+    t_fit = np.asarray(t_fit)
```
Is this necessary? I think it is fine to just insist that inputs are array-like. If you think it is important, use `np.asanyarray` to pass on `Quantity`.

larrybradley replied:
I thought "array-like" could include things like lists (basically anything that can be passed to `np.asanyarray`/`np.asarray`/`np.array`), so I think `np.asanyarray` is needed in that case.

mhvk replied:
I think my main question was whether this deep in the code it is necessary to be able to handle all kinds of different inputs, or whether it is fine to only be able to handle items that behave like `ndarray`.

jakevdp replied:
I would expect everything this deep to be an ndarray.

mhvk commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+    t_fit = np.asarray(t_fit)
+    assert t.ndim == 1
+    assert t_fit.ndim == 1
+    assert np.isscalar(frequency)
```
I would simply remove all the assert statements; fine if the code crashes on wrong input (and perhaps there is a way to use it to give multidimensional input). But if you keep it, do raise useful exceptions.

jakevdp replied:
OK  – @mwcraig mentioned something similar. I'll go through and change all the asserts to checks that raise an appropriate ``ValueError``

mhvk commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+
+    if center_data:
+        w = dy ** -2.0
+        y_mean = np.dot(y, w) / w.sum()
```
`np.dot` is not Quantity-safe, but here the alternative is, if anything, more readable:
```
y_mean = (y * w).sum() / w.sum()
```

jakevdp replied:
These low-level routines do not accept quantities.

mhvk commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+    X : ndarray, shape=(n_times, n_parameters)
+        The design matrix, where n_parameters = bool(bias) + 2 * nterms
+    """
+    t = np.asarray(t)
```
Like above, I think one should be able to *assume* input is array-like, so would remove it.  If you prefer to have it, do use `np.asanyarray`.

mhvk commented on astropy/stats/lombscargle/implementations/mle.py:
```diff
@@ -0,0 +1,98 @@
+    """
+    t = np.asarray(t)
+    assert t.ndim == 1
+    assert np.isscalar(frequency)
```
These should be tests with exceptions, or just be removed.

mhvk commented:
@jakevdp - I started commenting but am realizing the conversion to quantify-safe behaviour is not quite trivial. As I'm running out of time, for now please ignore those comments. I'll try to make a PR to yours that does it, but realise I should really start from a higher level.

mhvk commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,255 @@
+    available_methods = available_methods()
+
+    def __init__(self, t, y, dy=None, fit_bias=True, center_data=True):
+        self.t = t
```
My sense would be to do any `np.asanyarray` here, so that in the rest of the code you can safely assume that the inputs are arrays (or possibly `None` for `dy` if that is important).

I think one might as well do the broadcasting here too (which does `asanyarray` internally if `subok=True`):
```
from astropy.utilts.compat.numpy import broadcast_arrays

if dy is None:
    self.dy = None
    self.t, self.y = broadcast_arrays(t, y, subok=True)
else:
    self.t, self.y, self.dy = broadcast_arrays(t, y, dy, subok=True)
    # not strictly needed, but probably good.
    if hasattr(y, 'unit') and not (y[0]/dy[0]).unit.is_equivalent(u.dimensionless_unscaled):
        raise u.UnitsError('y and dy must have the same unit.')
if self.t.ndim != 1:
    raise ValueError('can only handle one-dimensional input.')
```

jakevdp replied:
Good idea – that would catch errors in the inputs earlier, which would probably be useful.

(FWIW, the reason I defaulted to this way is that I'm used to writing code for scikit-learn, where the policy is to have ``__init__`` do nothing but store the arguments by name)

mhvk commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,255 @@
+        frequency : ndarray or Quantity
+            The heuristically-determined optimal frequency bin
+        """
+        t = np.asanyarray(self.t)
```
If we go with the above, this can be removed.

mhvk commented on astropy/stats/lombscargle/implementations/chi2_impl.py:
```diff
@@ -0,0 +1,80 @@
+        dy = 1
+
+    t, y, dy = np.broadcast_arrays(t, y, dy)
+    frequency = np.asarray(frequency)
```
This will definitely go wrong with units, if the unit of the frequency is not simply the inverse of that of the times (e.g., `mHz` and `s`).

jakevdp replied:
I chose to have all the units taken care of in the wrapper functions, so that all these lower-level functions can expect raw numpy arrays.

You might reply that I should pass units all the way through and let the math happen with them. I initially tried this but it led to errors – I can't recall what exactly they were at the moment.

mhvk replied:
@jakevdp - before you change too much, how about my general comment about quantities and units, to do that all in `LombScargle`, and let that also take care of converting things to `array`.  With that, one could just remove all these broadcasts and asserts.

jakevdp replied:
Those are good ideas. I'm going to start on that now.

Also realized that some tests are in confusing places, because a while ago I reorganized the code but not the tests. Will take care of that as well.

mhvk commented:
@jakevdp - Looking closer, I think the underlying implementations are very much not `Quantity` safe, and I think they would fail if one passed in, e.g., frequency in `mHz` and time in `s`. In the end, it might be nice  to ensure all code is OK, but I think that is a drawn-out process with very little actual benefit, since all the user really sees is the `LombScargle` class. So I think it makes more sense to ensure the  class does appropriate conversions in `autopower`, `power` and `model`. I'd also suggest that those methods do any validation that is required, so that the implementations can assume dimensions, etc., are OK.

Does this make sense?

mhvk commented on astropy/stats/lombscargle/_lombscargle.py:
```diff
@@ -0,0 +1,255 @@
+    >>> LombScargle(t, y).power(freq)
+    array([ 0.0204304 ,  0.01393845,  0.35552682,  0.01358029,  0.03083737])
+
+    If the inputs are astropy Quantities with units, the units will be
```
My personal preference would be to just give an example that uses `Quantity`, since this is the astropy standard, but note that plain arrays are also OK. Ideally, use an example where unit conversion is done, say, using times in seconds, but checking at frequencies in `mHz`.

jakevdp commented:
@mhvk 

> Does this make sense?

That's exactly how I designed things here :smile:

jakevdp commented:
Overall comment as I read through all the individual ones: high-level routines are all quantity-safe. Low-level routines are not. This is by design: initially I tried passing quantities through all the low-level routines, and it caused a lot of strange corner-case issues.

mhvk commented:
Sorry for the to & fro: let's first agree that the way to go is to keep the low-level routines quantity-free (for now, maybe forever; lots of numpy routines like `np.dot` are not subclass-safe).

But that does mean you need to do proper unit conversion in `power`, etc....  I can help with that if you wish.

jakevdp commented:
All the proper unit conversion is already done in ``power``, etc, and the unit tests confirm it's working. The only one that's missing is ``model``, and I'm just about finished with that :smile:

mhvk commented on astropy/stats/lombscargle/tests/test_lombscargle.py:
```diff
@@ -0,0 +1,161 @@
+
+@pytest.mark.parametrize('method', LombScargle.available_methods)
+@pytest.mark.parametrize('t_unit', [units.second, units.day])
+@pytest.mark.parametrize('frequency_unit', [units.Hz, 1. / units.second])
```
While in principle this checks units, it doesn't actually check the result is correct!

In particular, since you are always evaluating at the same frequency, the results for `t_unit=u.s` should be different from that for `t_unit=u.day` (which I don't think will be the case currently).

mhvk replied:
Sorry, I see I missed your new input validation, which looks good, and should ensure this passes. I think it may still be a good idea to check the output is as expected, though!

mhvk commented on astropy/stats/lombscargle/tests/test_lombscargle.py:
```diff
@@ -0,0 +1,161 @@
+    frequency = np.linspace(0.5, 1.5, 10) * frequency_unit
+    PLS = lombscargle(t, y, frequency=frequency,
+                      fit_bias=False, method=method)
+    assert_equal(PLS.unit, units.dimensionless_unscaled)
```
For the unweighted case, I think the power spectrum should have units equal to those of the input squared, no? (E.g., if I use magnitudes, the powers should be `mag**2`.)

jakevdp replied:
PLS is equivalent to the chi-squared fit, which is unitless if you're taking uncertainties into account.

When uncertainties are not specified, it's a bit more tricky. Leaving uncertainties out of the chi-square completely does give units of mag**2, but I tend to think of no uncertainties as being one-and-the-same as uniform uncertainties, in which case the result is unitless.

mhvk replied:
Agreed on the case where uncertainties are given, but I think without them my tendency would be to think of the power as amplitude square of a sine or cosine fit, where the amplitude surely has the units of the data. My own sense is that that is least confusing: it ensures that `PLS(t, y.to(u.mag)) == PLS(t, y.to(u.mmag))` (which I think we would agree should hold!)

jakevdp replied:
Actually, should have said "one minus the chi-squared fit"

jakevdp replied:
That's one huge misconception: the periodogram does **not** reflect the amplitude of the Fourier fit.

jakevdp replied:
Rather, it reflects how well a sinusoid fits the data at each frequency.

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+    if t.ndim != 1:
+        raise ValueError("Input times & data must be one-dimensional")
+
+    has_units = any(isinstance(arr, units.Quantity)
```
My tendency would be to write this as `any(hasattr(arr, 'unit'))` -- but then I love ducktyping... (not important)

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+            if not y.unit.is_equivalent(dy.unit):
+                raise ValueError("Units of y not equivalent to units of dy")
+            dy = units.Quantity(dy, unit=y.unit)
+    else:
```
I think this is the place to set the power unit:
```
    power_unit = units.dimensionless_unscaled
else:
    power_unit = y.unit ** 2

jakevdp replied:
See the above comment. The power is always dimensionless.

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+
+        if frequency is not None:
+            frequency = units.Quantity(frequency)
+            if not t.unit.is_equivalent(1. / frequency.unit):
```
Since `Quantity` converts anyway, may be more logical as a try/except?
```
try:
    t = units.Quantity(t, unit=frequency.unit**(-1), copy=False)
except units.UnitConversionError:
    raise units.UnitConversionError("units of frequency not equi...
```
(`UnitConversionError` is subclassed from `ValueError` so someone can still catch it with that).

Aside: is there a particular reason to convert the time rather than the frequency?

jakevdp replied:
No, I can't think of any reason to favor one over the other

mhvk replied:
I think I'd convert frequency in that case, since `t` is always given and thus seems somewhat more "primary".

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+        if frequency is not None:
+            frequency = units.Quantity(frequency)
+            if not t.unit.is_equivalent(1. / frequency.unit):
+                raise ValueError("Units of frequency not equivalent to "
```
Do use `units.UnitConversionError` even if you keep the `if` statement

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+    if has_units:
+        power_unit = units.dimensionless_unscaled
+
+        t = units.Quantity(t)
```
Add `copy=False` to these initialisations, so that if something is already OK, it doesn't get unnecessarily copied (this makes it similar to `np.asarray`).  Of course, the moment conversions are needed, copies will get made.

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+        if dy is not None:
+            dy = np.asarray(dy)
+
+    def get_unit(val):
```
This can just be `getattr(val, 'unit', 1)`

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+                 'frequency': 1. / get_unit(t),
+                 'power': power_unit}
+
+    def unit_strip(arr):
```
This could just be `getattr(arr, 'value', arr)`

jakevdp replied:
In the case of a scalar, your version returns a scalar value rather than a 0-dimensional array and this breaks things down the line.

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,322 @@
+    unit_dict = {'t': get_unit(t),
+                 'y': get_unit(y),
+                 'dy': get_unit(y),
+                 'frequency': 1. / get_unit(t),
```
One could just have `get_unit(frequency)`, no? You already ensured the units were consistent.

jakevdp replied:
Frequency can be None, and we want the correct units in that case.

mhvk replied:
OK, makes sense.

mhvk commented:
@jakevdp - I think you may be reacting faster than I can comment!

jakevdp commented:
A few responses to @mwcraig 

> The assume_regular_frequency option shows up in a few places. I'm having trouble imagining cases where setting it to True would be the right thing to do....

The check takes a bit of time, so I thought I'd give users the option to turn it off at their own risk.

> ... I don't think the narrative documents explicitly say that the frequencies should not be angular frequencies...

Good point – I'll add a note on this to the narrative docs

> ... perhaps center_data could be removed altogether?

There's one situation in which you definitiely do not want to center the data: that's when you're using this to estimate the power of a window function. In that case you pass the algorithm an array of ones which you want it to use directly. If they're centered, the result is NaN across the board. That was one **huge** frustration I had with scipy's implementation before I wrote the gatspy code.

MSeifert04 commented:
Sorry for this comment but just to bring it to your attention, @jakevdp, you committed 7 times in the last 2 hours and one build takes approximatly 1 hour. This will block Travis builds for other PRs.

Would it be possible to test locally and only push after you did several commits (not after every commit)? Or commit and push with [skip ci] in the commit message? This will trigger the test builds less often and give other PRs some of the ressources.

I don't mean you shouldn't start the tests here but maybe less frequently.

mhvk commented on astropy/stats/lombscargle/implementations/chi2_impl.py:
```diff
@@ -0,0 +1,83 @@
+
+    p = np.array([compute_power(f) for f in frequency])
+
+    if normalization == 'unnormalized':
```
I thought I would just move the discussion to where an actual calculation is done. So, here, we have computed `p`. Looking at the code above, one has calculated a design matrix and solved it for amplitudes, which have the units of `y`, so the powers have units of `y.unit**2`. Now, if you normalize below by `chi2_ref`, you do get dimensionless, and I realize this is the default (and most useful case). But if you do not, then the unit here stays the same.

jakevdp replied:
The amplitutes are

```python
amp = np.linalg.solve(XTX, XTy)
```

Here we compute

```python
p = np.dot(XTy.T, amp)
```

Note that above, we start by scaling ``y`` by the uncertainty, so all the matrices are dimensionless.

jakevdp replied:
In particular, it's easy to see above that ``chi2ref`` is dimensionless. So if the result of dividing ``chi2`` by this is dimensionless as you say, then ``chi2`` must be dimensionless as well.

jakevdp replied:
Sorry – I made a mistake above. The matrices are not all dimensionless.

As written here, ``X`` has units ``1/y.unit``, ``XTX`` has units ``1/y.unit ** 2``, and ``XTy`` has units ``1/y.unit``.

This gives ``amp = np.linalg.solve(XTX, XTy)`` units of ``y.unit``, as expected for an amplitude. The  power is ``XTy * amp``, which is dimensionless.

It helped me to stare a bit at equation (12) of http://arxiv.org/pdf/1502.01344v1.pdf

mhvk replied:
I agree with the normalized case, but not for the unnormalized and unweighted case. Of course, this does boil down to what one assumes implicitly for weights. You seem to take `dy = 1 * y.unit`, while I think I prefer for unweighted to be truly unweighted, i.e., `dy` is ignored (or taken to be `1` without a unit).

Sorry to keep on about this so much, but my involvement with the `units` package has taught me how useful it is to be really obnoxious about units, as it helps a *lot* to catch mistakes.

E.g., assume someone wants to do the normalization outside of `LombScargle`, maybe using a different way of calculating the variance of `y` (which is what `chi2_ref` is), then we have to make sure the units end up being OK.

jakevdp replied:
Yes, that's where I started from somewhere above: I think of the absence of errors as equivalent to the assumption that all errors are equal.

At least we're in agreement that in the presence of uncertainties, the unnormalized result should be unitless, right? As far as the API goes, it would strike me as overly confusing to have the function return a unitless quantity if the data has uncertainties, and a unit-ful quantity otherwise.

mhvk replied:
Yes, I agree that in the presence of uncertainties, the output is unitless, and so it would be if one normalizes.

But for the unnormalized, unweighted case, it clearly was confusing to me to think that the output should never have a unit, so there is a counterexample :smiley_cat:  

I still think it is nicer to keep the units for this case, so that, as I wrote before, with quantities one has that `LS(t, y, normalize=False) == LS(t, y.to(0.001 * y.unit), normalize=False)`. I think this is especially important since it is now clear that this equality *will* hold if I put in any `dy` or if I choose `normalize=True`.

jakevdp replied:
OK – so you're proposing that if ``dy=None``, we treat it as "no uncertainties", so that the unnormalized periodogram has units of ``y.unit ** 2`` (the normalized periodogram is still unitless). If ``dy`` is specified, then we check that its units match ``y`` and return a unitless result.

Is that correct?

mhvk replied:
Yes, that's what to me seems most natural.  This in part since giving a known constant uncertainty is easy, while by assuming it is `1 * y.unit`, it becomes difficult to get the "no uncertainty at all" case.

jakevdp replied:
If so, I'm fine with that.

jakevdp replied:
I'll also add a section in the docs about these subtleties. In the meantime, we can brace for the inevitable mailing list question of how to get a unit-ful periodogram with data uncertainties. :smile: (I suspect the only answer is "you can't").

mhvk replied:
OK, sounds good.  Sorry for jumping in without fully having understood the code; could have gotten here quite a bit faster if I had!

jakevdp replied:
No worries – thanks for helping!

jakevdp commented:
Sorry – didn't realize multiple commits was an issue.

pllim commented:
I cancelled the Travis builds since there is still active discussions going on. Adding `[skip ci]` to commit until ready for testing might  be helpful here since everyone is on commit frenzy prior to feature freeze. :smile:

jakevdp commented:
Will do – thanks!

eteq commented on docs/stats/index.rst:
```diff
@@ -13,9 +13,17 @@ in astronomy and astropy.
+Most tools are fairly self-contained, and include relevant examples in
 their docstrings.
 
+Documentation
```
In most of the other doc sections, the "more detail" section like this is titled "Using <packagename>".  See, e.g. [the coordinates index page](https://github.com/astropy/astropy/blob/master/docs/coordinates/index.rst). I think this should be changed to 
```
Using `astropy.stats` 
=====================
```

eteq commented:
FWIW (harkening back a few comments), I think this should go into `astropy.stats` now once it's working even if #4846 is not resolved in a timely manner. It certainly fits in the realm of "standard algorithms that are in some mannerastro-specific" which is generally how the subpackage has been treated in the past.

Also, @jakevdp, the changelog needs to be updated (``CHANGES.rst``) - this certainly goes under the "new features" section under v1.2.

jakevdp commented:
I just pushed a commit renaming ``normalization='unnormalized'`` to ``normalization='psd'``, to make clear that this option results in the same normalization as the Fourier PSD. I also added a section to the documentation that discusses this.

jakevdp commented:
Also changed ``normalization='normalized'`` to ``normalization='standard'``

eteq commented:
Looks like there's a remaining failure coming from the fact that one of the documentation examples requires scipy, which is an "optional" dependency for Astropy.  There's an easy fix for this, @jakevdp - you can add the ``.. doctest-requires:: scipy`` directive. (See [number 4 in this section](http://docs.astropy.org/en/stable/development/testguide.html#skipping-doctests) for more on the topic)

crawfordsm commented:
@jakevdp Do you want to include a note and reference to your paper describing the code?  I think it would also be okay to include in that note something about citing that paper if this code is used.

jakevdp commented:
I've referenced it in the narrative docs. Do you think I should add a note in the doc string as well?

crawfordsm commented:
Yes, I think a note about that in the doc strings would be appropriate and be a place that would be harder to miss.

jakevdp commented:
A question for folks on naming: currently I have a ``fit_bias=True`` argument. Do you think it would be more clear to change this to ``floating_mean=True`` or perhaps ``generalized=True``? I'm open to any of those that would make sense.

jakevdp commented:
(I'm leaning slightly toward ``floating_mean``)

mhvk commented:
How about `fit_mean`?

mhvk commented on astropy/stats/lombscargle/core.py:
```diff
@@ -0,0 +1,338 @@
+        frequency : ndarray or Quantity
+            The heuristically-determined optimal frequency bin
+        """
+        t = np.asanyarray(self.t)
```
This can just be `t = self.t`, right? `broadcast_arrays` will have ensured this is an array.

jakevdp replied:
Yep

mhvk commented on astropy/stats/lombscargle/implementations/chi2_impl.py:
```diff
@@ -0,0 +1,88 @@
+    .. [3] Scargle, J.D. 1982, ApJ 263:835-853
+    """
+    if dy is None:
+        dy = 1
```
In principle, since we know `t`, `y` are validated arrays here, this could just be
```
dy = broadcast_to(1., y.shape)
```
(need to import this above)
With that, the next line can go.  But really, also fine to keep as is; overhead will be small and this way the routine can be used on its own more easily.

mhvk commented on astropy/stats/lombscargle/implementations/chi2_impl.py:
```diff
@@ -0,0 +1,88 @@
+
+    p = np.array([compute_power(f) for f in frequency])
+
+    if normalization == 'psd':
```
This part is common to all routines, so perhaps you want to take it out, and put it in `mle` or so? (Not essential)

jakevdp replied:
I kept it separate because each method computes the normalization differently. Comparing them all in the unit tests gives me more confidence that I've computed it correctly.

mhvk commented on astropy/stats/lombscargle/implementations/cython_impl.pyx:
```diff
@@ -0,0 +1,252 @@
+            S2 += 2 * w * sin_omega_t * cos_omega_t
+            C2 += w * (1 - 2 * sin_omega_t * sin_omega_t)
+
+        tau = 0.5 * atan(S2 / C2) / omega[i]
```
Should this use `atan2` to ensure one picks the right quadrant?

jakevdp replied:
I think in this case the quadrant won't matter, because tau is only used directly as a phase within a sine or cosine that is eventually squared.

jakevdp replied:
Although now that I think of it ``atan2`` will correctly handle the case of zero in the denominator, so it may be worth changing.

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,227 @@
+    return frequency[0], frequency[1] - frequency[0], len(frequency)
+
+
+def _is_regular(frequency, assume_regular_frequency=False):
```
Minor detail, but why the `assume_regular_frequency` argument, instead of just using it below?

jakevdp replied:
no reason – I'll change it.

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,227 @@
+    elif len(frequency) == 1:
+        return True
+    else:
+        diff = frequency[1:] - frequency[:-1]
```
this can just be
```
diff = np.diff(frequency)
```

mhvk commented on astropy/stats/lombscargle/implementations/main.py:
```diff
@@ -0,0 +1,227 @@
+    methods = available_methods()
+    fast_method_ok = ('fast' in methods)
+    prefer_fast = (fast_method_ok and len(frequency) > 200
+                   and _is_regular(frequency, assume_regular_frequency))
```
So `and (assume_regular_frequency or _is_regular(frequency))`
(`_is_regular` will not get executed if `assume_regular_frequency` holds, since python short-circuits these expressions).

mhvk commented on astropy/stats/lombscargle/implementations/scipy_impl.py:
```diff
@@ -0,0 +1,74 @@
+
+import numpy as np
+
+try:
```
I'd move this inside `lombscargle_scipy`, so that `scipy` only gets imported when it is actually used. (I'd also change `ValueError` to `ImportError`, i.e., just adjust the error message)

mhvk commented on astropy/stats/lombscargle/tests/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+This packages contains affiliated package tests.
```
This won't be an affiliated package!  So, adapt or just remove the docstring.

mhvk commented:
@jakevdp - very minor comments only, as you can see. Also just to be sure, astropy testing does run all your tests, right (also in `implementation`)?

I think if it had been up to me, I would have assumed all input is properly set up for each implementation, rather than testing again. For me, the clarity of code (and my tolerance for code just breaking if it gets improper input) would outweigh the possible argument that one might want to use those routines on their own. But, really, this is one of those things that are up to whoever does the actual work, so I'll just note it, and let you decide.

jakevdp commented:
Thanks for the comments. Regarding testing – I'd be happy to make all the tests use the primary interface if you think that would be clearer and/or easier to maintain.

mhvk commented:
I just wanted to be sure that the tests were run, which you can probably tell from running them locally (and we can check once travis completes).
But if it is not too much effort, yes, I think it is easiest to put the two test files in the same place.

jakevdp commented:
Is there any reason tests within the package would not be run?

mhvk commented:
> Is there any reason tests within the package would not be run?

I think they are run, but I don't know how it is done well enough to be sure -- but anyway, you can just check output of `./setup.py test -P stats` or we can look at the travis logs once they are available.

jakevdp commented:
I know they run & pass locally – I confirm that every time before I push. I'm just curious if there's something else I should be looking out for in general.

mhvk commented:
OK, then all is OK. I just looked at where other astropy test files are located, and while most are simply under `astropy.*.tests`, some are located in their module. I think it is fine to keep as is.

jakevdp commented:
OK @mhvk I responded to your comments, refactored the tests, and changed ``fit_bias`` to ``fit_mean`` throughout. Let me know if you see anything else!

crawfordsm commented:
I can confirm the tests are all run from setup.py and astropy.test, so I think where they are is fine.

astrofrog commented:
@jakevdp - if this PR is otherwise ready, can you amend the last commit to remove ``skip ci``, so that we know what the latest CI status is?

jakevdp commented:
Pushed a doc commit so we'll get a CI test.

astrofrog commented:
@jakevdp - one small change to make sure the coverage doesn't go down so much - we need to exclude the generated C code (from Cython) from the coverage testing (we do this for all generated C code as opposed to directly written C code). You need to edit ``.travis.yml`` and at the end of the file, in the line starting

```
cpp-coveralls -E ".*convolution.*" -E "...
```

you need to add a new exclusion:

```
-E ".*cython_impl.*"
```

The AppVeyor failure seems to be unrelated, I'll look into that separately.

crawfordsm commented:
Once the coverage is cleared up, I'm happy to merge this.   I did ask for an independent review and their opinion was that it was providing consistent results with other packages while being impressed with the overall performance.   

So thank you @jakevdp for the great submission!

jakevdp commented:
Test failure is in a timing utility, unrelated to this PR.

astrofrog commented:
@crawfordsm - this is good to go, please merge if you think it's ready!

mhvk commented:
@jakevdp - I only noticed now, but the travis run shows that many tests fail with numpy dev -- this is not immediately important since it may actually be a bug in numpy, but could you have a look at the log and see if you can figure out what goes wrong? See https://travis-ci.org/astropy/astropy/jobs/130002187

MSeifert04 commented:
@mhvk Do you happen to know if any recent changes in numpy-dev changed the return of reduce-like-ufuncs (i.e. `sum`) or the behaviour of numpy scalars? From the Exceptions in the linked log it seems like they now return python scalars instead of numpy scalars or the numpy scalars now behave like python scalars.

The difference:
```
>>> import numpy as np
>>> 1.0 / np.array([-1, 0, 1]).sum()  # numpy scalar numpy 1.10 and 1.11
RuntimeWarning: divide by zero encountered in long_scalars
inf

>>> 1.0 / 0.0 # python scalars
ZeroDivisionError: float division by zero
```

jakevdp commented:
Thanks @mhvk – I think https://github.com/astropy/astropy/pull/4950 should fix it.
